airbyte-integrations/connectors/source-salesforce/unit_tests/integration/test_bulk_stream.py

```diff
@@ -65,29 +71,225 @@ def tearDown(self) -> None:
+        )
+        self._mock_delete_job(_JOB_ID)
+
+        output = read(_STREAM_NAME, SyncMode.full_refresh, self._config)
+
+        assert len(output.records) == 2
+
+    def test_given_job_creation_have_transient_error_when_read_then_sync_properly(self):
+        given_stream(self._http_mocker, _BASE_URL, _STREAM_NAME, SalesforceDescribeResponseBuilder().field(_A_FIELD_NAME))
+        self._http_mocker.post(
+            self._make_full_job_request([_A_FIELD_NAME]),
+            [
+                _RETRYABLE_RESPONSE,
+                JobCreateResponseBuilder().with_id(_JOB_ID).build(),
+            ],
+        )
+        self._http_mocker.get(
+            HttpRequest(f"{_BASE_URL}/jobs/query/{_JOB_ID}"),
+            JobInfoResponseBuilder().with_id(_JOB_ID).with_state("JobComplete").build(),
+        )
+        self._http_mocker.get(
+            HttpRequest(f"{_BASE_URL}/jobs/query/{_JOB_ID}/results"),
+            HttpResponse(f"{_A_FIELD_NAME}\nfield_value"),
+        )
+        self._mock_delete_job(_JOB_ID)
+
+        output = read(_STREAM_NAME, SyncMode.full_refresh, self._config)
+
+        assert len(output.errors) == 0
+        assert len(output.records) == 1
+
+    def test_given_bulk_restrictions_when_read_then_switch_to_standard(self):
+        given_stream(self._http_mocker, _BASE_URL, _STREAM_NAME, SalesforceDescribeResponseBuilder().field(_A_FIELD_NAME))
+        self._http_mocker.post(
+            self._make_full_job_request([_A_FIELD_NAME]),
+            [
+                HttpResponse("[{}]", 403),
+                JobCreateResponseBuilder().with_id(_JOB_ID).build(),
+            ],
+        )
+        self._http_mocker.get(
+            create_standard_http_request(_STREAM_NAME, [_A_FIELD_NAME]),
+            create_standard_http_response([_A_FIELD_NAME]),
+        )
+
+        output = read(_STREAM_NAME, SyncMode.full_refresh, self._config)
+
+        assert len(output.records) == 1
+
+    def test_given_non_transient_error_on_job_creation_when_read_then_fail_sync(self):
+        given_stream(self._http_mocker, _BASE_URL, _STREAM_NAME, SalesforceDescribeResponseBuilder().field(_A_FIELD_NAME))
+        self._http_mocker.post(
+            self._make_full_job_request([_A_FIELD_NAME]),
+            HttpResponse(json.dumps([{"errorCode": "API_ERROR", "message": "Implementation restriction... <can't complete the error message as I can't reproduce this issue>"}]), 400),
+        )
+
+        output = read(_STREAM_NAME, SyncMode.full_refresh, self._config)
+
+        assert output.get_stream_statuses(_STREAM_NAME)[-1] == AirbyteStreamStatus.INCOMPLETE
+
+    def test_given_job_is_aborted_when_read_then_fail_sync(self):
+        given_stream(self._http_mocker, _BASE_URL, _STREAM_NAME, SalesforceDescribeResponseBuilder().field(_A_FIELD_NAME))
+        self._http_mocker.post(
+            self._make_full_job_request([_A_FIELD_NAME]),
+            JobCreateResponseBuilder().with_id(_JOB_ID).build(),
+        )
```

Comment:
nit: self._stream_and_mock_post()? Maybe it's worth it as these lines are duplicated 3 or 4 times.

Reply:
I see this but they are not necessarily related so I wouldn't put them both in one method (one is for the stream discover and the other is for the job creation. I don't mind extracting the `self._http_mocker.post(...)` into something like `def given_job_creation(fields, created_job_id)` if you'd like

Reply:
No, it is fine; as I said, it was just nit stuff but I think you are right. Resolving.